### DISCUSSION TITLE
Performance issue while adding constraints

### DISCUSSION THREAD

**Discussion by Ervin66** [coin-or/pulp#436] [Q&A] [answered]
Dear all,

when googling for ways to speed up the constraints generation in a lot of old SO threads it is mentioned that using "+=" is awfully slow but I have yet to find what should be used instead. There was also mention of a patch for the "+=" operator has this been implemented?

If this has not been patched could this someone provide an example of what should be done instead? I think this questions has been asked by many but there hasn't been any satisfactory answer? [example](https://stackoverflow.com/questions/26643421/pulp-very-slow-when-adding-many-constraints)

Thank you all in advance

**Comment by stumitchell** [comment]
lpSum() and a list comprehension

so
c = lpSum([x for x in X])

**Reply by Ervin66** [reply]
but isn't that just summing over the element within lpSum?

if we take this example from the documentation how would this work?
```python
for b in Bars:
    prob += lpSum([vars[w][b] for w in Warehouses])>=demand[b]
```

**Comment by pchtsp** [comment]
Do you have a specific minimum reproducible example in python we could check? Stuart's answer is correct. It just may be that the inefficiencies are coming from somewhere else in the code (or in the structure of the list comprehension).

In some very rare occasions, there are some very niche alternatives that can work better. That's why I'm asking for an example. And why there's no "single definitive solution" (that is, besides `lpSum` + `list comprehension`)

**Reply by Ervin66** [reply]
I am not sure whether I understand the list comprehension in this case. I already do a lsit comprehension in order to sum over the elements but how could I use it to generate all the constraints and then only add them later?
```python
for b in Bars:
    prob += lpSum([vars[w][b] for w in Warehouses])>=demand[b]
```
How would you iterate over the Bars list in this case in order to add all constraints at once?

**Reply by stumitchell** [reply]
@Ervin66 you are currently doing things right.
The suggestion you refer to is there because often people come to pulp and generate code like the following
```py
for b in Bars:
    c = LpAffineExpression()
    for w in Warehouses:
        c += vars[w][b]
    prob +=c >=demand[b]
```

in this case the `+=` in this line `c += vars[w][b]` makes the code very slow

**Comment by stumitchell** [comment, marked as answer]
Pulp uses the `+=` operator to add constraints to `LpProblem`s this is idiomatic usage.

However often people come from other libraries and languages and use `+=` to generate single constraints like the following
```py
for b in Bars:
    c = LpAffineExpression()
    for w in Warehouses:
        c += vars[w][b]
    prob +=c >=demand[b]
```
in this case the `+=` in this line `c += vars[w][b]` makes the code very slow

**Reply by Ervin66** [reply]
thanks for the reply and help! :)

But for my objective function I create a lpSum element for each individual summation and once they are all created I add to the model at once which does speed the creation of the objective function example:
```python
c1 = lpSum[(x[i][j]*c[i][j],
                 for i in I,
                 for j in J])
c2 = lpSum([Y[i]*fc[i],
                  for i in I])

prob += c1 + c2

```
So there is no equivalent for constraints to be all added simultaneously?

**Comment by stumitchell** [comment]
LpProblem.extend()

will do what you want

https://github.com/coin-or/pulp/blob/b99ff24891123fae928c6571d2eccd3e073eb385/pulp/pulp.py#L1681

Stuart Mitchell
PhD Engineering Science
Extraordinary Freelance Programmer and Optimisation Guru
www.stuartmitchell.com

**Reply by Ervin66** [reply]
I will look into it, thank you stu
